File: blue/src/package_neon_multi_node_blue/tools.py
"""The steps and every template spec, the port of io.github.getcolors.neon-multi-node.tools."""

from __future__ import annotations

import json
import math
import os
from decimal import Decimal
from pathlib import Path

import package_neon_blue
from blue import tofu
from blue.ansible import ansible_with_spec, parse_recap
from blue.cli import stage_dir
from blue.runtime import runtime
from blue.scaffold import PRESERVE_JINJA_DELIMITERS, content_spec
from colors_compute.orchestration import orchestrate
from colors_compute.planning import plan_deployment
from blue.scaffold import scaffold

from . import ssh_config, topology, validate, storage
from .utils import clj_str as _s

infrastructure_tool = "neon-multi-node-infrastructure"
dns_tool = "neon-multi-node-dns"
ansible_tool = "neon-multi-node-ansible"
ansible_local_tool = "neon-multi-node-ansible-local"
ROOT = Path(__file__).parent / "resources"

# Pinned Neon supplies generic Ansible config; application resources are owned here.
NEON_ROOT = Path(package_neon_blue.__file__).parent / "resources"

template_opts = PRESERVE_JINJA_DELIMITERS


def tool_dir(opts: dict, tool: str) -> str:
    return stage_dir(opts, tool, default_profile="neon-multi-node")


def template(path: str, file: str) -> dict:
    name = f"tools/{path}/{file}"
    return {"name": name, "content": (ROOT / name).read_text()}


def neon_template(path: str, file: str) -> dict:
    name = f"tools/{path}/{file}"
    return {"name": f"neon/{name}", "content": (NEON_ROOT / name).read_text()}


def spec(source: dict, target: str, data: dict) -> dict:
    return {"template": source, "target": target, "data": data, "opts": template_opts}


def raw_spec(target: str, content: str) -> dict:
    return content_spec(target, content)


def credential_env(opts: dict, *slots: str) -> dict[str, str] | None:
    merged: dict[str, str] = {}
    for slot in [*slots, "provider-backend"]:
        merged.update(validate.tofu_env(opts, slot))
    result = {}
    for key, env_var in merged.items():
        value = _s(opts.get(key))
        if value:
            result[env_var] = value
    return result or None


def backend_credential_env(opts: dict) -> dict[str, str] | None:
    return credential_env(opts)


# ------------------------------------------------------------------- json


def _java_double(value: float) -> str:
    """`Double.toString`, which is what Cheshire writes for a double: plain
    decimal with at least one fractional digit between 1e-3 and 1e7, and
    `d.dddE<n>` computerized scientific notation outside that range."""
    if math.isnan(value):
        return "NaN"
    if math.isinf(value):
        return "Infinity" if value > 0 else "-Infinity"
    if value == 0:
        return "-0.0" if math.copysign(1.0, value) < 0 else "0.0"
    sign = "-" if value < 0 else ""
    digits, exponent = Decimal(repr(abs(value))).as_tuple()[1:]
    text = "".join(str(d) for d in digits).rstrip("0") or "0"
    # the decimal point sits after `point` digits of `text`
    point = len(digits) + exponent
    if 1e-3 <= abs(value) < 1e7:
        if point <= 0:
            return f"{sign}0.{'0' * (-point)}{text}"
        if point >= len(text):
            return f"{sign}{text}{'0' * (point - len(text))}.0"
        return f"{sign}{text[:point]}.{text[point:]}"
    mantissa = text[0] + "." + (text[1:] or "0")
    return f"{sign}{mantissa}E{point - 1}"


def _json_scalar(value) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        return _java_double(value)
    return json.dumps(str(value), ensure_ascii=False)


def _pretty(value, indent: int = 0) -> str:
    """Cheshire's pretty JSON, byte for byte — Green's artifact contract. In
    insertion order: `tofu.constructs_json` sorts keys, which is right for a
    Terraform document and wrong for the documents this package writes
    itself, where the order is the caller's."""
    if isinstance(value, (list, tuple)):
        if not value:
            return "[ ]"
        return "[ " + ", ".join(_pretty(item, indent) for item in value) + " ]"
    if isinstance(value, dict):
        if not value:
            return "{ }"
        pad = " " * (indent + 2)
        body = ",\n".join(f"{pad}{json.dumps(str(k), ensure_ascii=False)} : {_pretty(v, indent + 2)}"
                          for k, v in value.items())
        return "{\n" + body + "\n" + " " * indent + "}"
    return _json_scalar(value)


# ------------------------------------------------------------- compute output


def hosts(opts: dict) -> list[dict]:
    """The host list for every stage after compute: the recorded cluster
    under `colors-compute/cluster` on a real run, library plans on a build (see
    `topology.hosts`)."""
    return topology.hosts(opts)


# ---------------------------------------------------------------- compute

def http_sources(opts):return {'source':'explicit','ranges':[]}
def ranges_checksum(values):return 'none'

async def infrastructure_step(opts):
    resolved = http_sources(opts)
    if opts.get('blue/event') == 'create' and not opts.get('blue/dry-run') and resolved['source'] == 'fallback':
        return {**opts, 'blue/exit': 1, 'blue/err': 'Cloudflare ingress ranges unavailable; refusing stale fallback'}
    requirements = topology.requirements(opts)
    try:
        if opts.get('blue/event') == 'build' or opts.get('blue/dry-run'):
            result = plan_deployment(opts, topology.topology(opts), requirements)
            specs = []
            root = Path(opts['workdir']) / opts['profile'] / 'compute'
            for name, document in result['documents']['shared'].items():
                specs.append(raw_spec(str(root / 'shared' / name), json.dumps(document, indent=2, sort_keys=True) + '\n'))
            for node_id, documents in result['documents']['nodes'].items():
                for name, document in documents.items():
                    specs.append(raw_spec(str(root / 'nodes' / node_id / name), json.dumps(document, indent=2, sort_keys=True) + '\n'))
            specs.append(raw_spec(str(root / 'http-sources.json'), _pretty({'origin': resolved['source'], 'checksum': ranges_checksum(resolved['ranges']), 'ranges': resolved['ranges']})))
            scaffold(opts, specs)
        else:
            result = await orchestrate(opts, topology.topology(opts), requirements, {**os.environ, **storage.aws_env(opts)})
        if result['status'] not in ('planned', 'ready', 'destroyed'):
            return {**opts, 'blue/exit': 1, 'blue/err': '\n'.join(result.get('errors', [])) or 'compute lifecycle refused'}
        output = {**opts, 'blue/exit': 0}
        if 'cluster' in result:
            output['colors-compute/cluster'] = result['cluster']
            output['colors-compute/shared'] = result.get('shared', {})
        path = result.get('key', {}).get('private_key_path')
        if path:
            output['ssh-private-key-path'] = path.replace('$HOME', '/home/build-placeholder') if result['status'] == 'planned' else path
        return output
    except (ValueError, KeyError):
        return {**opts, 'blue/exit': 1, 'blue/err': 'invalid compute deployment requirements'}


# ------------------------------------------------------------------- dns


def zone_id() -> str:
    return "${data.cloudflare_zone.zone.id}"


def dns_json(opts: dict, app_ip: str | None) -> str:
    """DNS-only public A record; PostgreSQL cannot use the HTTP proxy."""
    return tofu.constructs_json([
        tofu.construct("resource", "cloudflare_dns_record", "neon-multi-node", {
            "zone_id": zone_id(),
            "name": opts.get("neon-host"),
            "type": "A",
            "content": app_ip,
            "ttl": 1,
            "proxied": bool(opts.get("cloudflare-proxied")),
        }),
    ])


async def dns_step(opts: dict) -> dict:
    dir = tool_dir(opts, dns_tool)
    app = topology.host_of(hosts(opts), "compute") or {}
    specs = [spec(template("dns", "main.tf"), f"{dir}/main.tf", opts),
             raw_spec(f"{dir}/record.tf.json", dns_json(opts, app.get("ip")))]
    return await tofu.tofu_with_spec(
        opts, specs, dir=dir, env=credential_env(opts, "provider-dns"))


# ------------------------------------------------------- ssh config (local)


def ansible_local_data(opts: dict) -> dict:
    """Only what a `build` genuinely knows. Addresses are run-time facts and
    reach the play as extra-vars instead, so the rendered playbook carries no
    IP and is identical on every workstation (SSH Config Standard §6)."""
    return {**opts,
            "ssh-keygen": validate.keygen(opts),
            "ssh-config-identity-file": ssh_config.identity_file(opts) if validate.keygen(opts) else opts.get("ssh-private-key-path", "")}


def ansible_local_specs(opts: dict) -> list[dict]:
    dir = tool_dir(opts, ansible_local_tool)
    data = ansible_local_data(opts)
    # ansible.cfg and the inventory are the dependency's, unchanged; the play
    # is this package's own because it writes six stanzas, not one.
    return [spec(neon_template("ansible-local", "ansible.cfg"), f"{dir}/ansible.cfg", data),
            spec(neon_template("ansible-local", "inventory.ini"), f"{dir}/inventory.ini", data),
            spec(template("ansible-local", "main.yml"), f"{dir}/main.yml", data)]


def ssh_config_hosts(opts, hosts_):
    app = topology.host_of(hosts_, 'compute')
    return [{'name': ssh_config.host_alias(opts), 'ip': app['ip'], 'user': app['user']},
            *[{'name': ssh_config.machine_alias(opts, h), 'ip': h['ip'], 'user': h['user']} for h in hosts_]]


async def ansible_local_step(opts: dict) -> dict:
    """Write or remove the `~/.ssh/config` block. The same playbook serves
    both events; `block_state` is what distinguishes them."""
    dir = tool_dir(opts, ansible_local_tool)
    delete = opts.get("blue/event") == "delete"
    return await ansible_with_spec(
        opts, ansible_local_specs(opts),
        dir=dir, inventory="inventory.ini",
        playbooks={"create": "main.yml", "delete": "main.yml"},
        extra_vars={"host_alias": ssh_config.host_alias(opts),
                    "ssh_hosts": ssh_config_hosts(opts, hosts(opts)),
                    "block_state": "absent" if delete else "present"})



def inventory(opts,nodes):
    children={}
    for role in ['compute','pageserver','safekeeper']:
        children[role]={'hosts':{h['node_id']:{'ansible_host':h['ip'],'ansible_user':h['user'],'vpc_ip':h['vpc_ip'],'role':role,'ordinal':h['index']} for h in sorted(nodes,key=lambda x:x['node_id']) if h['role']==role}}
    return _pretty({'all':{'children':children}})

def ansible_data(opts):
    return {**{k:v for k,v in opts.items() if k!='neon-multi-node/storage-credentials'},'domain':opts['neon-host'],'ssh-keygen':validate.keygen(opts)}

ANSIBLE_FILES=['site.yml','cleanup.yml','rehearsal.yml','runtime.py','bootstrap.sh','scramgen.py','compute-spec.json','acceptance.py','renew-tls.sh','rehearse-member.yml']

def ansible_specs(opts):
    directory=tool_dir(opts,ansible_tool);data=ansible_data(opts)
    return [spec(neon_template('ansible','ansible.cfg'),directory+'/ansible.cfg',data),
            raw_spec(directory+'/inventory.json',inventory(opts,hosts(opts))),
            *[spec(template('ansible',name),directory+'/'+name,data) for name in ANSIBLE_FILES]]

def cleanup_proof(opts,output):
    import re
    expected={n['node_id'] for n in hosts(opts)}
    recap=parse_recap(output)
    reported=set(re.findall(r'(?m)^\s*"msg": "WRITERS_STOPPED ([a-z0-9-]+) PASS independent absence of Neon containers and writer processes"\s*$',output))
    if not (len(expected)==5 and expected==reported and expected==set(recap) and all(c['ok']>0 and all(c.get(k,-1)==0 for k in ['failed','unreachable','rescued','ignored']) for c in recap.values())):
        raise RuntimeError('writer stop proof incomplete; storage deletion refused')
    return expected

async def run_play(opts,playbook,credentials=False):
    rendered={**scaffold({**opts,'blue/event':'build'},ansible_specs(opts)),'blue/event':opts.get('blue/event')}
    if opts.get('blue/event')=='build':return {**rendered,'blue/exit':0}
    env={'ANSIBLE_HOST_KEY_CHECKING':'True','ANSIBLE_SSH_ARGS':'-o StrictHostKeyChecking=accept-new'}
    if credentials:env.update(storage.credential_env(opts))
    result=await runtime.exec(['ansible-playbook','-i','inventory.json',playbook],cwd=tool_dir(opts,ansible_tool),env=env,timeout_ms=7200000)
    if result.exit:return {**rendered,'blue/exit':1,'blue/err':'Ansible failed: '+result.out+result.err}
    if playbook=='cleanup.yml':
        try:
            stopped=cleanup_proof(opts,result.out)
            target=Path(opts['workdir'])/opts['profile']/'writer-stop-evidence.txt';target.parent.mkdir(parents=True,exist_ok=True);target.write_text(result.out)
            return {**rendered,'blue/exit':0,'ansible/recap':parse_recap(result.out),'neon-multi-node/writers-stopped':stopped}
        except Exception as e:return {**rendered,'blue/exit':1,'blue/err':str(e)}
    return {**rendered,'blue/exit':0,'ansible/recap':parse_recap(result.out)}

async def ansible_step(opts):
    if opts.get('blue/event')=='delete' and opts.get('colors-compute/cluster') is None:return {**opts,'blue/exit':0}
    return await run_play(opts,'cleanup.yml' if opts.get('blue/event')=='delete' else 'site.yml',opts.get('blue/event')=='create')
async def rehearsal_step(opts):return await run_play(opts,'rehearsal.yml',True)
async def acceptance_step(opts):
    if opts.get('blue/event')=='build' or opts.get('blue/dry-run'):return {**opts,'blue/exit':0}
    cfg={k:opts[k] for k in ['profile','neon-host','neon-role','neon-database']}
    result=await runtime.exec(['python3','acceptance.py',json.dumps(cfg)],cwd=tool_dir(opts,ansible_tool),timeout_ms=600000)
    return {**opts,'blue/exit':0 if result.exit==0 else 1,'blue/err':result.out+result.err}
async def describe_step(opts):
    try:
        for node in hosts(opts):
            alias=opts['profile']+'-'+node['node_id']
            result=await runtime.exec(['ssh','-o','BatchMode=yes',alias,'sudo','-n','python3','/opt/neon/runtime.py','status'])
            rows=[json.loads(line) for line in result.out.splitlines() if line.strip()]
            expected=2 if node['role']=='pageserver' else 1
            if result.exit or len(rows)!=expected or not all(row['State']=='running' and row.get('Health')!='unhealthy' for row in rows):raise RuntimeError('unhealthy role: '+node['node_id'])
            print('PASS healthy '+node['node_id'])
        return {**opts,'blue/exit':0}
    except Exception as e:return {**opts,'blue/exit':1,'blue/err':str(e)}

File: red/src/tools.ts
import * as ansible from 'red/ansible';
import {stageDir} from 'red/cli';
import {scaffold,PRESERVE_JINJA_DELIMITERS,contentSpec,type Spec} from 'red/scaffold';
import * as tofu from 'red/tofu';
import {runtime} from 'red/runtime';
import type {Opts} from 'red/workflow';
import {orchestrate,plan_deployment} from 'colors-compute-red';
import {mkdirSync,writeFileSync} from 'node:fs';
import {dirname,join} from 'node:path';
import * as topology from './topology.ts';
import * as storage from './storage.ts';
import * as sshConfig from './ssh-config.ts';
import * as validate from './validate.ts';
import {resources} from './resources.ts';
export const infrastructureTool='neon-multi-node-infrastructure',dnsTool='neon-multi-node-dns',ansibleTool='neon-multi-node-ansible',ansibleLocalTool='neon-multi-node-ansible-local';
export const toolDir=(opts:Opts,tool:string)=>stageDir(opts,tool,{defaultProfile:'neon-multi-node'});
export const hosts=topology.hosts;
const spec=(path:string,target:string,data:Opts):Spec=>({template:{name:path,content:resources[path]!},target,data,opts:PRESERVE_JINJA_DELIMITERS});
export function credentialEnv(opts:Opts,...slots:string[]):Record<string,string>{
 const mapping=Object.assign({},...[...slots,'provider-backend'].map(slot=>validate.tofuEnv(opts,slot)));
 return Object.fromEntries(Object.entries(mapping).filter(([k])=>validate.s(opts[k])).map(([k,v])=>[String(v),String(opts[k])]));
}
// Cheshire layout for package-owned JSON, including insertion-ordered inventory fields.
export function pretty(value:any,indent=0):string{
 if(Array.isArray(value))return value.length?'[ '+value.map(x=>pretty(x,indent)).join(', ')+' ]':'[ ]';
 if(value!==null&&typeof value==='object'){const entries=Object.entries(value);return entries.length?'{'+'\n'+entries.map(([k,v])=>' '.repeat(indent+2)+JSON.stringify(k)+' : '+pretty(v,indent+2)).join(',\n')+'\n'+' '.repeat(indent)+'}':'{ }';}
 return JSON.stringify(value??null);
}
const sorted=(value:any):any=>Array.isArray(value)?value.map(sorted):value&&typeof value==='object'?Object.fromEntries(Object.keys(value).sort().map(k=>[k,sorted(value[k])])):value;
export async function infrastructureStep(opts:Opts,deps:any={}):Promise<Opts>{
 try{
  const planning=opts['red/event']==='build'||opts['red/dry-run'];
  const result:any=planning?plan_deployment(opts,topology.topology(opts),topology.requirements(opts)):await (deps.orchestrate??orchestrate)(opts,topology.topology(opts),topology.requirements(opts),{...process.env,...storage.awsEnv(opts)});
  if(planning){const root=join(opts.workdir,opts.profile,'compute');const stages:any[]=[['shared',result.documents.shared],...Object.entries(result.documents.nodes).map(([id,docs])=>['nodes/'+id,docs])];for(const[stage,docs]of stages){mkdirSync(join(root,stage),{recursive:true});for(const[name,doc]of Object.entries(docs))writeFileSync(join(root,stage,name),JSON.stringify(sorted(doc),null,2)+'\n');}writeFileSync(join(root,'http-sources.json'),pretty({origin:'explicit',checksum:'none',ranges:[]}));}
  if(!['planned','ready','destroyed'].includes(result.status))return {...opts,'red/exit':1,'red/err':'compute lifecycle refused; legacy monolithic state requires explicit migration'};
  return {...opts,'red/exit':0,...(result.cluster?{'colors-compute/cluster':result.cluster,'colors-compute/shared':result.shared}:{}),...(result.key?.private_key_path?{'ssh-private-key-path':planning?result.key.private_key_path.replace('$HOME','/home/build-placeholder'):result.key.private_key_path}:{})};
 }catch(e){return {...opts,'red/exit':1,'red/err':'compute deployment failed: '+(e as Error).message};}
}
export const dnsJson=(opts:Opts,ip:string)=>tofu.constructsJson([tofu.construct('resource','cloudflare_dns_record','neon-multi-node',{zone_id:'${data.cloudflare_zone.zone.id}',name:opts['neon-host'],type:'A',content:ip,ttl:1,proxied:Boolean(opts['cloudflare-proxied'])})]);
export async function dnsStep(opts:Opts):Promise<Opts>{const dir=toolDir(opts,dnsTool);return tofu.tofuWithSpec(opts,[spec('dns/main.tf',dir+'/main.tf',opts),contentSpec(dir+'/record.tf.json',dnsJson(opts,topology.hostOf(hosts(opts),'compute')!.ip))],{dir,env:credentialEnv(opts,'provider-dns')});}
export function ansibleLocalData(opts:Opts):Opts{return {...opts,'ssh-keygen':validate.keygen(opts),'ssh-config-identity-file':validate.keygen(opts)?sshConfig.identityFile(opts):opts['ssh-private-key-path']??''};}
export function ansibleLocalSpecs(opts:Opts):Spec[]{const dir=toolDir(opts,ansibleLocalTool),data=ansibleLocalData(opts);return ['ansible.cfg','inventory.ini','main.yml'].map(name=>spec('ansible-local/'+name,dir+'/'+name,data));}
export function sshConfigHosts(opts:Opts,list:topology.Host[]){return [{...topology.hostOf(list,'compute'),name:opts.profile},...list.map(h=>({...h,name:sshConfig.machineAlias(opts,h)}))];}
export async function ansibleLocalStep(opts:Opts):Promise<Opts>{return ansible.ansibleWithSpec(opts,{dir:toolDir(opts,ansibleLocalTool),inventory:'inventory.ini',playbooks:{create:'main.yml',delete:'main.yml'},extraVars:{host_alias:sshConfig.hostAlias(opts),ssh_hosts:sshConfigHosts(opts,hosts(opts)),block_state:opts['red/event']==='delete'?'absent':'present'}},ansibleLocalSpecs(opts));}
export function inventory(_opts:Opts,list:topology.Host[]):string{
 const children=Object.fromEntries(['compute','pageserver','safekeeper'].map(role=>[role,{hosts:Object.fromEntries(list.filter(h=>h.role===role).sort((a,b)=>a.node_id.localeCompare(b.node_id)).map(h=>[h.node_id,{ansible_host:h.ip,ansible_user:h.user,vpc_ip:h.vpc_ip,role:h.role,ordinal:h.index}]))}]));
 return pretty({all:{children}});
}
export function ansibleData(opts:Opts):Opts{const{'neon-multi-node/storage-credentials':_secret,...rest}=opts;return {...rest,domain:opts['neon-host'],'ssh-keygen':validate.keygen(opts)};}
export const ansibleFiles=['site.yml','cleanup.yml','rehearsal.yml','runtime.py','bootstrap.sh','scramgen.py','compute-spec.json','acceptance.py','renew-tls.sh','rehearse-member.yml'];
export function ansibleSpecs(opts:Opts):Spec[]{const dir=toolDir(opts,ansibleTool),data=ansibleData(opts);return [spec('ansible/ansible.cfg',dir+'/ansible.cfg',data),contentSpec(dir+'/inventory.json',inventory(opts,hosts(opts))),...ansibleFiles.map(name=>spec('ansible/'+name,dir+'/'+name,data))];}
export function cleanupProof(opts:Opts,output:string):Set<string>{
 const expected=hosts(opts).map(h=>h.node_id).sort(),recap=ansible.parseRecap(output);
 const reported=[...output.matchAll(/^\s*"msg": "WRITERS_STOPPED ([a-z0-9-]+) PASS independent absence of Neon containers and writer processes"\s*$/gm)].map(m=>m[1]!);
 const equal=(a:string[],b:string[])=>JSON.stringify([...new Set(a)].sort())===JSON.stringify([...new Set(b)].sort());
 if(expected.length!==5||!equal(expected,reported)||!equal(expected,Object.keys(recap))||!Object.values(recap).every((row:any)=>row.ok>0&&['failed','unreachable','rescued','ignored'].every(k=>row[k]===0)))throw Error('writer stop proof incomplete; storage deletion refused');
 return new Set(expected);
}
export async function runPlay(opts:Opts,playbook:string,credentials:boolean,deps:any={}):Promise<Opts>{
 if(opts['red/event']==='build')return {...scaffold(opts,ansibleSpecs(opts)),'red/exit':0};
 const rendered={...scaffold({...opts,'red/event':'build'},ansibleSpecs(opts)),'red/event':opts['red/event']};
 const env={ANSIBLE_HOST_KEY_CHECKING:'True',ANSIBLE_SSH_ARGS:'-o StrictHostKeyChecking=accept-new',...(credentials?storage.credentialEnv(opts):{})};
 const result=await(deps.exec??runtime.exec)(['ansible-playbook','-i','inventory.json',playbook],{cwd:toolDir(opts,ansibleTool),env,timeoutMs:7200000});
 if(result.exit)return {...rendered,'red/exit':1,'red/err':'Ansible failed: '+result.out+result.err};
 if(playbook==='cleanup.yml')try{const stopped=cleanupProof(opts,result.out),target=join(opts.workdir,opts.profile,'writer-stop-evidence.txt');mkdirSync(dirname(target),{recursive:true});writeFileSync(target,result.out);return {...rendered,'red/exit':0,'ansible/recap':ansible.parseRecap(result.out),'neon-multi-node/writers-stopped':stopped};}catch(e){return {...rendered,'red/exit':1,'red/err':(e as Error).message};}
 return {...rendered,'red/exit':0,'ansible/recap':ansible.parseRecap(result.out)};
}
export const ansibleStep=(opts:Opts):Promise<Opts>=>opts['red/event']==='delete'&&!opts['colors-compute/cluster']?Promise.resolve({...opts,'red/exit':0}):runPlay(opts,opts['red/event']==='delete'?'cleanup.yml':'site.yml',opts['red/event']==='create');
export const rehearsalStep=(opts:Opts)=>runPlay(opts,'rehearsal.yml',true);
export async function acceptanceStep(opts:Opts):Promise<Opts>{if(opts['red/event']==='build'||opts['red/dry-run'])return {...opts,'red/exit':0};const cfg=Object.fromEntries(['profile','neon-host','neon-role','neon-database'].map(k=>[k,opts[k]]));const result=await runtime.exec(['python3','acceptance.py',JSON.stringify(cfg)],{cwd:toolDir(opts,ansibleTool),timeoutMs:600000});return {...opts,'red/exit':result.exit===0?0:1,'red/err':result.out+result.err};}
export async function describeStep(opts:Opts):Promise<Opts>{try{for(const node of hosts(opts)){const result=await runtime.exec(['ssh','-o','BatchMode=yes',opts.profile+'-'+node.node_id,'sudo','-n','python3','/opt/neon/runtime.py','status'],{timeoutMs:60000});const rows=result.out.split(/\r?\n/).filter(x=>x.trim()).map(x=>JSON.parse(x));if(result.exit||rows.length!==(node.role==='pageserver'?2:1)||rows.some(r=>r.State!=='running'||r.Health==='unhealthy'))throw Error('container health check failed');runtime.log('PASS healthy '+node.node_id);}return {...opts,'red/exit':0};}catch{return {...opts,'red/exit':1,'red/err':'one or more role services unavailable'};}}
